lib/good_job/active_job_extensions/concurrency.rb

```diff
@@ -0,0 +1,41 @@
+module GoodJob
+  module ActiveJobExtensions
+    module Concurrency
+      extend ActiveSupport::Concern
+
+      ConcurrencyExceededError = Class.new(StandardError)
+
+      included do
+        retry_on GoodJob::ActiveJobExtensions::Concurrency::ConcurrencyExceededError, attempts: Float::INFINITY, wait: :exponentially_longer
+      end
+
+      class_methods do
+        def good_job_enqueue_exclusively_with(limit: 1)
+          before_enqueue do |job|
+            next unless job.respond_to? :good_job_concurrency_key
+            next if $ERROR_INFO.is_a? GoodJob::ActiveJobExtensions::Concurrency::ConcurrencyExceededError
+
+            GoodJob::Job.new.with_advisory_lock(key: job.good_job_concurrency_key, function: "pg_advisory_lock") do
+              # TODO: Why is `unscoped` necessary? Nested scope is bleeding into subsequent query?
+              enqueue_concurrency = GoodJob::Job.unscoped.where(concurrency_key: job.good_job_concurrency_key).unfinished.count
+              # The job has not yet been enqueued, so check if adding it will go over the limit
+              throw :abort if enqueue_concurrency + 1 > limit
```

Comment:
Does this create a race condition that needs to be accounted for? Say the enqueue_concurrency limit is 1, and the actual enqueue concurrency goes from 1 to 0 between the above two lines of code.

Reply:
I think the race condition is addresses because this is wrapped with a `pg_advisory_lock` on the key, which should function as a database-wide mutex. 

But It's also gonna introduce latency because the PG connection will wait until it can acquire an advisory lock before running the code. If done on the webserver when under heavy load and there's a database statement timeout... that could be bad. I _hope_ that jobs that are likely to need enqueue limits will be triggered from a backend process (rake task, upcoming cron feature, etc).